2011-03-10  Michael Tautschnig  <mt@debian.org>

	* setup-storage/Parser.pm: Build an additional list of logical volumes to make
		on-disk order of logical volumes consistent with disk_config.
	* setup-storage/Commands.pm: Implement this order.
Index: trunk/lib/setup-storage/Parser.pm
===================================================================
--- trunk.orig/lib/setup-storage/Parser.pm
+++ trunk/lib/setup-storage/Parser.pm	
@@ -807,6 +807,8 @@
           # make sure, $2 has not been defined already
           defined ($FAI::configs{$FAI::device}{volumes}{$2}{size}{range}) and 
             die "Logical volume $2 has been defined already.\n";
+          # add to ordered list
+          push @{ $FAI::configs{$FAI::device}{ordered_lv_list} }, $2;
           # initialise the new hash
           defined($FAI::configs{$FAI::device}{volumes}{$2}) or
             $FAI::configs{$FAI::device}{volumes}{$2} = {};
@@ -860,6 +862,8 @@
             $FAI::configs{$FAI::device}{volumes} = {};
           # initialise the list of physical devices
           $FAI::configs{$FAI::device}{devices} = ();
+          # initialise the ordered list of volumes
+          $FAI::configs{$FAI::device}{ordered_lv_list} = ();
           # init device tree
           $FAI::dev_children{$FAI::device} = ();
           # the rule must not return undef
@@ -1143,6 +1147,12 @@
       }
     } elsif ($config =~ /^VG_(.+)$/) {
       next if ($1 eq "--ANY--");
+      (scalar(keys %{ $FAI::configs{$config}{volumes} }) ==
+        scalar(@{ $FAI::configs{$config}{ordered_lv_list} })) or
+        &FAI::internal_error("Inconsistent LV lists - missing entries");
+      defined($FAI::configs{$config}{volumes}{$_}) or
+        &FAI::internal_error("Inconsistent LV lists - missing entries")
+        foreach (@{ $FAI::configs{$config}{ordered_lv_list} });
       foreach my $p (keys %{ $FAI::configs{$config}{volumes} }) {
         my $this_mp = $FAI::configs{$config}{volumes}{$p}{mountpoint};
         next if ($this_mp eq "-");
Index: trunk/lib/setup-storage/Commands.pm
===================================================================
--- trunk.orig/lib/setup-storage/Commands.pm
+++ trunk/lib/setup-storage/Commands.pm	
@@ -522,7 +522,7 @@
   my $vg = $1; # the actual volume group
 
   # now create or resize the configured logical volumes
-  foreach my $lv (keys %{ $FAI::configs{$config}{volumes} }) {
+  foreach my $lv (@{ $FAI::configs{$config}{ordered_lv_list} }) {
     # reference to the size of the current logical volume
     my $lv_size = (\%FAI::configs)->{$config}->{volumes}->{$lv}->{size};
     # skip preserved partitions, but ensure that they exist
